### PR TITLE
Allow dynamic clients to be used in downstream OIDC flows

### PR DESCRIPTION
**Pull request by cfryanr**
The dynamically configured OIDCClient CRs actually work in the downstream Supervisor OIDC flows.

**Release note**:

```release-note
TODO as part of the overall release note for the dynamic clients feature
```

### CODE REVIEW

File: internal/oidc/clientregistry/clientregistry.go [hunk 1]
```diff
@@ -11,47 +11,109 @@ import (
...
+// Client implements the base, OIDC, and response_mode client interfaces of Fosite.
 var (
 	_ fosite.Client              = (*Client)(nil)
 	_ fosite.OpenIDConnectClient = (*Client)(nil)
 	_ fosite.ResponseModeClient  = (*Client)(nil)
 )
 
-// StaticClientManager is a fosite.ClientManager with statically-defined clients.
-type StaticClientManager struct{}
+func (c Client) GetResponseModes() []fosite.ResponseModeType {
```
Comment (enj) [inline internal/oidc/clientregistry/clientregistry.go:40]:
```suggestion
func (c *Client) GetResponseModes() []fosite.ResponseModeType {
```

?

File: internal/oidc/clientregistry/clientregistry.go [hunk 2]
```diff
@@ -11,47 +11,109 @@ import (
...
+// use the plain error's text as that error's debug message (see client_authentication.go in fosite).
+func (m *ClientManager) GetClient(ctx context.Context, id string) (fosite.Client, error) {
+	if id == PinnipedCLIClientID {
+		// Return the static client. No lookups needed.
 		return PinnipedCLI(), nil
-	default:
+	}
+
+	// Try to look up an OIDCClient with the given client ID (which will be the Name of the OIDCClient).
+	oidcClient, err := m.oidcClientsClient.Get(ctx, id, v1.GetOptions{})
```
Comment (enj) [inline internal/oidc/clientregistry/clientregistry.go:82]: Before doing the lookup, I would check if the name has the required prefix.

File: internal/supervisor/server/server.go [hunk 3]
```diff
@@ -439,6 +439,7 @@ func runSupervisor(ctx context.Context, podInfo *downward.PodInfo, cfg *supervis
 		dynamicUpstreamIDPProvider,
 		&secretCache,
 		clientWithoutLeaderElection.Kubernetes.CoreV1().Secrets(serverInstallationNamespace), // writes to kube storage are allowed for non-leaders
+		clientWithoutLeaderElection.PinnipedSupervisor.ConfigV1alpha1().OIDCClients(serverInstallationNamespace),
```
Comment (enj) [inline internal/supervisor/server/server.go:442]:
```suggestion
		client.PinnipedSupervisor.ConfigV1alpha1().OIDCClients(serverInstallationNamespace),
```

**Review by enj — COMMENTED** [review]
I have not looked at the tests.

File: internal/oidc/clientregistry/clientregistry.go (hunk 1 above)
Comment (cfryanr) [reply internal/oidc/clientregistry/clientregistry.go]: Updated in 88f611d31a07382565ca78485980023ac8ef2145.

File: internal/oidc/clientregistry/clientregistry.go (hunk 2 above)
Comment (cfryanr) [reply internal/oidc/clientregistry/clientregistry.go:93]: Updated in 88f611d31a07382565ca78485980023ac8ef2145.

File: internal/supervisor/server/server.go (hunk 3 above)
Comment (cfryanr) [reply internal/supervisor/server/server.go]: Updated in 88f611d31a07382565ca78485980023ac8ef2145.

**Pull request merged**